// @adpt/cloud/dist/src/k8s/ServiceAccount.js
"use strict";
/*
 * Copyright 2021 Unbounded Systems, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.isServiceAccountProps = exports.serviceAccountResourceInfo = exports.ServiceAccount = void 0;
const tslib_1 = require("tslib");
const core_1 = tslib_1.__importStar(require("@adpt/core"));
const common_1 = require("./common");
const k8s_observer_1 = require("./k8s_observer");
const manifest_support_1 = require("./manifest_support");
const Resource_1 = require("./Resource");
const defaultProps = {
    metadata: {}
};
/**
 * k8s ServiceAccount resource
 *
 * @param propsIn - props for ServiceAccount as specifed in {@link k8s.ServiceAccountProps}
 *
 * @public
 */
function ServiceAccount(propsIn) {
    const _a = propsIn, { handle, key, config, metadata } = _a, rest = tslib_1.__rest(_a, ["handle", "key", "config", "metadata"]);
    return core_1.default.createElement(Resource_1.Resource, Object.assign({ key: key, kind: "ServiceAccount", config: config, metadata: metadata }, rest));
}
exports.ServiceAccount = ServiceAccount;
// TODO: The "as any" is a workaround for an api-extractor bug. See issue #185.
ServiceAccount.defaultProps = defaultProps;
/** @internal */
exports.serviceAccountResourceInfo = {
    kind: "ServiceAccount",
    deployedWhen: () => true,
    statusQuery: async (props, observe, buildData) => {
        const obs = await observe(k8s_observer_1.K8sObserver, core_1.gql `
            query ($name: String!, $kubeconfig: JSON!, $namespace: String!) {
                withKubeconfig(kubeconfig: $kubeconfig) {
                    readCoreV1NamespacedServiceAccount(name: $name, namespace: $namespace) @all(depth: 100)
                }
            }`, {
            name: manifest_support_1.resourceIdToName(props.key, buildData.id, buildData.deployID),
            kubeconfig: props.config.kubeconfig,
            namespace: common_1.computeNamespaceFromMetadata(props.metadata)
        });
        return obs.withKubeconfig.readCoreV1NamespacedServiceAccount;
    },
    makeManifest: (manifest, elem, _deployID) => {
        const { automountServiceAccountToken, imagePullSecrets, secrets } = elem.props;
        return Object.assign(Object.assign({}, manifest), { automountServiceAccountToken,
            imagePullSecrets,
            secrets });
    }
};
/** @public */
function isServiceAccountProps(p) {
    return (p.kind === "ServiceAccount");
}
exports.isServiceAccountProps = isServiceAccountProps;
manifest_support_1.registerResourceKind(exports.serviceAccountResourceInfo);
