// @adpt/core/dist/src/reanimate/index.js
"use strict";
/*
 * Copyright 2018-2019 Unbounded Systems, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.reanimateDom = exports.registerObject = exports.registerConstructor = exports.reanimateUrn = exports.reanimate = exports.findMummyUrn = exports.findMummy = void 0;
var reanimate_1 = require("./reanimate");
Object.defineProperty(exports, "findMummy", { enumerable: true, get: function () { return reanimate_1.findMummy; } });
Object.defineProperty(exports, "findMummyUrn", { enumerable: true, get: function () { return reanimate_1.findMummyUrn; } });
Object.defineProperty(exports, "reanimate", { enumerable: true, get: function () { return reanimate_1.reanimate; } });
Object.defineProperty(exports, "reanimateUrn", { enumerable: true, get: function () { return reanimate_1.reanimateUrn; } });
Object.defineProperty(exports, "registerConstructor", { enumerable: true, get: function () { return reanimate_1.registerConstructor; } });
Object.defineProperty(exports, "registerObject", { enumerable: true, get: function () { return reanimate_1.registerObject; } });
var reanimate_dom_1 = require("./reanimate_dom");
Object.defineProperty(exports, "reanimateDom", { enumerable: true, get: function () { return reanimate_dom_1.reanimateDom; } });
